app/models/container_label_tag_mapping.rb

```diff
@@ -9,91 +9,91 @@ class ContainerLabelTagMapping < ApplicationRecord
+        # consider what will happen to previously created tags.
+        any_value.map do |tag_id|
+          {:category_tag_id   => tag_id,
+           :entry_name        => Classification.sanitize_name(value),
+           :entry_description => value}
+        end
       end
     end
   end
-  private_class_method :tags_for_name_type_value
+  private_class_method :map_name_type_value
 
-  # If this is an open ended any-value mapping, finds or creates a
-  # specific-value mapping to a specific tag.
-  def self.create_specific_value_mapping(name, type, value, category_tag)
-    new_tag = create_tag(name, value, category_tag)
-    new_mapping = create!(:labeled_resource_type => type, :label_name => name, :label_value => value,
-                          :tag => new_tag)
-    load_mapping_into_hash(new_mapping)
-    new_mapping
+  # Given a hash built by `map_*` methods, returns a Tag (creating if needed).
+  def self.find_or_create_tag(tag_hash)
+    if tag_hash[:tag_id]
+      Tag.find(tag_hash[:tag_id])
+    else
+      category = Tag.find(tag_hash[:category_tag_id]).classification
+      entry = category.find_entry_by_name(tag_hash[:entry_name])
+      unless entry
+        category.lock :exclusive do
+          begin
+            entry = category.add_entry(:name        => tag_hash[:entry_name],
+                                       :description => tag_hash[:entry_description])
+            entry.save!
+          rescue ActiveRecord::RecordInvalid
+            entry = category.find_entry_by_name(tag_hash[:entry_name])
+          end
+        end
+      end
+      entry.tag
+    end
   end
-  private_class_method :create_specific_value_mapping
 
-  def self.create_tag(name, value, category_tag)
-    category = category_tag.classification
-    unless category
-      category = Classification.create_category!(:description => "Kubernetes label '#{name}'",
-                                                 :read_only   => true,
-                                                 :tag         => category_tag)
-    end
+  def self.controls_tag?(tag)
+    return false unless tag.classification.try(:read_only) # never touch user-assignable tags.
+    tag_ids = [tag.id, tag.category.tag_id].uniq
+    where(:tag_id => tag_ids).any?
+  end
 
-    if value.empty?
-      entry_name = ':empty:' # ':' character won't occur in kubernetes values.
-      description = '<empty value>'
-    else
-      entry_name = Classification.sanitize_name(value)
-      description = value
+  # Assign/unassign mapping-controlled tags, preserving user-assigned tags.
+  def self.retag_entity(entity, tag_hashes)
+    mapped_tags = tag_hashes.map { |tag_hash| find_or_create_tag(tag_hash) }
+    existing_tags = entity.tags
+    Tagging.transaction do
+      (mapped_tags - existing_tags).each do |tag|
+        Tagging.create!(:taggable => entity, :tag => tag)
+      end
+      (existing_tags - mapped_tags).select { |tag| controls_tag?(tag) }.tap do |tags|
```

Comment:
I first tried the more symmetric `.each` with `Tagging.find_by(...).destroy` but was concerned find_by might return nil and felt doing `.try(:destroy)` is ugly.
This way if some taggings are not found, destroy_all just gets a smaller collection, there is no special case to consider.  It's also a bit more efficient.
(`delete_all` would be most efficient, not instantiating AR for callbacks; Tagging is a trivial model with no callbacks so it should work; but I felt that's an implementation detail this class shouldn't rely on.)

I'm now unsure how the Tagging could be not found.  These are mapping-conrolled tags, "no one else should be touching them".  Yeah, I've been burned by similar thinking before :-)
What do you say — leave like this, do `each ... .try(:destroy)`, or just `each ... .destroy` (save_inventory will catch and log exception if it does happen)?